### PR TITLE
[Delivers #97164178] restrict creation of proposals to users with GSA emails

### PR DESCRIPTION
**Pull request by afeld**
...if the `RESTRICT_ACCESS` environment variable is set to `true`.

### CODE REVIEW

**Comment by afeld** [comment]
I should add: we already have restriction on the viewing and editing of WorkOrders – this pull request optionally limits who can create them.

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -19,6 +19,7 @@ def authorize(record, query=nil, user=nil)
     user ||= @current_user
     record = self.authorizing_object(record)
     policy = Pundit.policy(user, record)
+    # TODO fall back to ProposalPolicy if there isn't a use-case-specific one
```
Comment (cmc333333) [inline app/controllers/application_controller.rb]: We do; see `authorizing_object`

File: app/controllers/use_case_controller.rb [hunk 2]
```diff
@@ -3,6 +3,7 @@
 # * permitted_params
 class UseCaseController < ApplicationController
   before_filter :authenticate_user!
+  before_filter ->{authorize self.model_class.new}, only: [:new, :create]
```
Comment (cmc333333) [inline app/controllers/use_case_controller.rb]: I don't think you need to authorize `self.model_class.new` separately due to `authorizing_object`

File: spec/policies/gsa18f/procurement_policy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+  permissions :can_create? do
+    with_feature 'RESTRICT_ACCESS' do
+      it "doesn't allow someone with a non-GSA email to create" do
+        user = User.new(email_address: 'intruder@some.com')
+        procurement = Gsa18f::Procurement.new
+        expect(subject).not_to permit(user, procurement)
+      end
+    end
+  end
```
Comment (cmc333333) [inline spec/policies/gsa18f/procurement_policy_spec.rb]: This covers the `RESTRICT_ACCESS=true` x non-gsa pair, but we probably want to test the three other permutations: (`RA=true`, gsa), (`RA=false`, non-gsa), (`RA=false`, gsa)

**Comment by cmc333333** [comment]
This _would_ be more generic if GSA wasn't hard coded. What do you think about a regex env variable?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (afeld) [inline app/controllers/application_controller.rb]: I think I had something wrong with my test that was causing a problem...*meh*.

File: app/controllers/use_case_controller.rb (hunk 2 above)
Comment (afeld) [inline app/controllers/use_case_controller.rb]: Meaning it should just be `Proposal.new`?

**Comment by afeld** [comment]
> This would be more generic if GSA wasn't hard coded.

I figure the restriction is going to vary by use case anyway (e.g. if we built one out for another agency), so not sure what benefit we'd get from making the details of the restriction configurable.

File: app/controllers/use_case_controller.rb (hunk 2 above)
Comment (cmc333333) [inline app/controllers/use_case_controller.rb]: I thought you could collapse this into the following line, but the object doesn't exist yet, my mistake.

Authorizing on an empty object still seems suspicious. Would this work: `authorize self.model_class`?

**Comment by cmc333333** [comment]
> I figure the restriction is going to vary by use case anyway
Sounds like a good candidate to be stored in the DB then ;) We can stick with this for now, but I'll throw another red flag here about hard coding values.

**Comment by afeld** [comment]
Updated!

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -34,6 +34,11 @@ def authorize(record, query=nil, user=nil)
 
   # Proposals can have special authorization parameters in their client_data
   def authorizing_object(record)
+    if record.is_a?(Class)
+      # use an instance
+      record = record.new
+    end
```
Comment (cmc333333) [inline app/controllers/application_controller.rb]: I'm not sure this is required, but it doesn't hurt

**Comment by cmc333333** [comment]
Thanks!

**Pull request merged**